dolfyn/adp/turbulence.py

```diff
@@ -0,0 +1,753 @@
+        are not actually vertical derivatives of the horizontal
+        velocity, then this is not the 'horizontal shear squared'.
+
+        See Also
+        --------
+        :math:`dudz`, :math:`dvdz`
+        """
+
+        tau2 = self.dudz(vel) ** 2 + self.dvdz(vel) ** 2
+        tau2.attrs['units'] = 's-2'
+        tau2.attrs['long_name'] = 'Horizontal Shear Squared'
+        tau2.attrs['standard_name'] = 'radial_sea_water_shear_squared'
+
+        return tau2
+
+    def calc_doppler_noise(self, psd, pct_fN=0.8):
+        """Calculate bias due to Doppler noise using the noise floor
+        of the velocity spectra.
+
+        Parameters
+        ----------
+        psd : xarray.DataArray (time, f)
+          The velocity spectra from a single depth bin (range), typically
+          in the mid-water range
+        pct_fN : float
+          Percent of Nyquist frequency to calculate characeristic frequency
+
+        Returns
+        -------
+        doppler_noise (xarray.DataArray): 
+          Doppler noise level in units of m/s
+
+        Notes
+        -----
+        Approximates bias from
+
+        .. :math: \\sigma^{2}_{noise} = N x f_{c}
+
+        where :math: `\\sigma_{noise}` is the bias due to Doppler noise,
+        `N` is the constant variance or spectral density, and `f_{c}`
+        is the characteristic frequency.
+
+        The characteristic frequency is then found as 
+
+        .. :math: f_{c} = pct_fN * (f_{s}/2)
+
+        where `f_{s}/2` is the Nyquist frequency.
+
+
+        Richard, Jean-Baptiste, et al. "Method for identification of Doppler noise 
+        levels in turbulent flow measurements dedicated to tidal energy." International 
+        Journal of Marine Energy 3 (2013): 52-64.
+
+        Thiébaut, Maxime, et al. "Investigating the flow dynamics and turbulence at a 
+        tidal-stream energy site in a highly energetic estuary." Renewable Energy 195 
+        (2022): 252-262.
+        """
+
+        if len(psd.shape) != 2:
+            raise Exception('PSD should be 2-dimensional (time, frequency)')
+
+        # Characteristic frequency set to 80% of Nyquist frequency
+        fN = self.fs/2
+        fc = pct_fN * fN
+
+        # Get units right
+        if psd.freq.units == "Hz":
+            f_range = slice(fc, fN)
+        else:
+            f_range = slice(2*np.pi*fc, 2*np.pi*fN)
+
+        # Noise floor
+        N2 = psd.sel(freq=f_range) * psd.freq.sel(freq=f_range)
+        noise_level = np.sqrt(N2.mean(dim='freq'))
+
+        out = xr.DataArray(
+            noise_level.values.astype('float32'),
+            dims=['time'],
+            attrs={'units': 'm s-1',
+                   'long_name': 'Doppler Noise Level',
+                   'description': 'Doppler noise level calculated '
+                   'from PSD white noise'})
+        return out
+
+    def calc_stress_4beam(self, ds, noise=None, orientation=None, beam_angle=25):
```

Comment:
The `beam_angle` can't be pulled from `ds` somewhere? It should be there somewhere...

Reply:
It's only stored in TRDI datasets, as far as I've found

Reply:
It's in the `dat.config` in my old datastructure [screenshot](https://user-images.githubusercontent.com/2273361/224578926-336fb816-2338-421d-9178-ef921a72b41d.png). I'm guessing this got dropped when we switched to xarray?  We really should be carrying `config` around.  Ideally this info should be stored in nested-dicts (and lists/tuples as appropriate), that can easily be accessed. However, storing that kind of info in xarray's attrs may be difficult? Maybe we can store everything in `dat.attrs['config_json_str']`, and then run `parse_nortek_config(dat.attrs['config_json_str'])` string whenever we actually need the data. With the `parse_...` function being built as-necessary?

Reply:
Hmmm, I thought I made sure everything in the config_json_str dictionary was captured in the dataset's global attributes.... I will double check that. A "parse_nortek_config" as you mention it is part of what the nortek readers are already doing.
Both nested-dicts and the json str are incompatible with the netcdf format, which was why I "exploded" the nested-dicts into a single dict that are now the dataset attributes. They work fine with xarray, but can't be saved into netcdf.